src/vcpkg/binarycaching.cpp

```diff
@@ -1255,6 +1255,162 @@ namespace
+                                 const Path& download_path,
+                                 MessageSink& sink) const
+        {
+            Command cmd = base_cmd(src, package_name, package_version, "download");
+            cmd.string_arg("--path").string_arg(download_path);
+            return run_az_artifacts_cmd(cmd, sink);
+        }
+
+        ExpectedL<Unit> publish(const AzureUpkgSource& src,
+                                StringView package_name,
+                                StringView package_version,
+                                const Path& package_dir,
+                                StringView description,
+                                MessageSink& sink) const
+        {
+            Command cmd = base_cmd(src, package_name, package_version, "publish");
+            cmd.string_arg("--description").string_arg(description).string_arg("--path").string_arg(package_dir);
+            return run_az_artifacts_cmd(cmd, sink);
+        }
+
+        ExpectedL<Unit> run_az_artifacts_cmd(const Command& cmd, MessageSink& sink) const
+        {
+            RedirectedProcessLaunchSettings show_in_debug_settings;
+            show_in_debug_settings.echo_in_debug = EchoInDebug::Show;
+            return cmd_execute_and_capture_output(cmd, show_in_debug_settings)
+                .then([&](ExitCodeAndOutput&& res) -> ExpectedL<Unit> {
+                    if (res.exit_code == 0)
+                    {
+                        return {Unit{}};
+                    }
+
+                    // az command line error message: Before you can run Azure DevOps commands, you need to
+                    // run the login command(az login if using AAD/MSA identity else az devops login if using PAT token)
+                    // to setup credentials.
+                    if (res.output.find("you need to run the login command") != std::string::npos)
+                    {
+                        sink.println(Color::warning,
+                                     msgFailedVendorAuthentication,
+                                     msg::vendor = "Universal Packages",
+                                     msg::url = "https://learn.microsoft.com/cli/azure/authenticate-azure-cli");
+                    }
+                    return LocalizedString::from_raw(std::move(res).output);
+                });
+        }
+        Path az_cli;
+    };
+
+    struct AzureUpkgPutBinaryProvider : public IWriteBinaryProvider
+    {
+        AzureUpkgPutBinaryProvider(const ToolCache& cache, MessageSink& sink, std::vector<AzureUpkgSource>&& sources)
+            : m_azure_tool(cache, sink), m_sources(sources)
+        {
+        }
+
+        size_t push_success(const BinaryPackageWriteInfo& request, MessageSink& msg_sink) override
+        {
+            size_t count_stored = 0;
+            auto ref = make_feedref(request, "");
+            std::string package_description = "Cached package for " + ref.id;
```

Comment:
I think it would be a good idea to have this be in `FeedReference` so that it can be tested rather than made up here.